### ISSUE TITLE
Komodo 9.0 problem with remote dir

### ISSUE THREAD

**Issue by johnnyamr** [Komodo/KomodoEdit#252]
Hi
I installed the latest version of komodo, keeping the version 8.5 (win7). When trying to connect to a remote account shows me the following error:
```
RemoteDir.prototype.connect[Exception... "Component returned failure code: 0x80004005 (NS_ERROR_FAILURE) [koISSHConnection.list]"  nsresult: "0x80004005 (NS_ERROR_FAILURE)"  location: "JS frame :: chrome://sites/content/servers.js :: this.init_rowsList/RemoteDir.prototype.connect :: line 317"  data: no]
```

The account that is trying to connect via SFTP and had no problems with the same account and settings in version 8.5. I tried uninstalling and installing both versions only version 9 and neither works. If someone has been the same, please please show me how it has solved or that may be the cause. Thanks.

Regards

**Comment by cgchoffman** [comment]
No issue for me connecting over SFTP.  Could you find/open/edit the following file.
``...Komodo IDE 9\lib\mozilla\components\koSFTP.py``, uncomment line 26, restart Komodo, try to connect again then attach the full logs from Help > Troubleshooting > View Log File here in a Gist or code section.

**Comment by johnnyamr** [comment]
Hi, this is what shows me the log file when trying to connect:

[2015-05-05 16:24:24,150] [DEBUG] koSFTPConnection: list: (refresh:True) path: '/home/httpd/html'
[2015-05-05 16:24:24,150] [DEBUG] koSFTPConnection: Checking that the SSH connection is alive
[2015-05-05 16:24:24,150] [DEBUG] koSFTPConnection: do_getPathInfo: Getting path info for: /
[2015-05-05 16:24:24,161] [DEBUG] koSFTPConnection: do_getParentPath: '/' -> '/'
[2015-05-05 16:24:24,161] [DEBUG] koSFTPConnection: do_getPathInfo: Getting path info for: /home/httpd/html
[2015-05-05 16:24:24,167] [DEBUG] koSFTPConnection: do_getParentPath: '/home/httpd/html' -> '/home/httpd'
[2015-05-05 16:24:24,167] [DEBUG] koSFTPConnection: list: Returned new rf_info for: '/home/httpd/html'
[2015-05-05 16:24:24,168] [DEBUG] koSFTPConnection: list: Creating new listing for directory: '/home/httpd/html'
ERROR:xpcom:Unhandled exception calling 'int8 * list(in wstring, in boolean, out retval nsISomething);'
Traceback (most recent call last):
  File "C:\Program Files\ActiveState Komodo Edit 9\lib\mozilla\python\xpcom\server\policy.py", line 300, in _CallMethod_
    return 0, func(*params)
  File "C:\Program Files\ActiveState Komodo Edit 9\lib\mozilla\python\komodo\remotefilelib.py", line 512, in list
    dirlist = self.do_getDirectoryList(path, cached_info)
  File "C:\Program Files\ActiveState Komodo Edit 9\lib\mozilla\components\koSFTP.py", line 278, in do_getDirectoryList
    info = self._sftp.listdir_attr(path)
  File "C:\Program Files\ActiveState Komodo Edit 9\lib\mozilla\python\komodo\paramiko\sftp_client.py", line 208, in listdir_attr
    filename = msg.get_text()
  File "C:\Program Files\ActiveState Komodo Edit 9\lib\mozilla\python\komodo\paramiko\message.py", line 199, in get_text
    return u(self.get_bytes(self.get_size()))
  File "C:\Program Files\ActiveState Komodo Edit 9\lib\mozilla\python\komodo\paramiko\py3compat.py", line 51, in u
    return s.decode(encoding)
  File "C:\Program Files\ActiveState Komodo Edit 9\lib\python\lib\encodings\utf_8.py", line 16, in decode
    return codecs.utf_8_decode(input, errors, True)
UnicodeDecodeError: 'utf8' codec can't decode byte 0xf1 in position 23: invalid continuation byte
[2015-05-05 16:24:24,220] [ERROR] console-logger: TypeError: properties is undefined (2) in chrome://sites/content/dinamicTree.js:150
Traceback (most recent call last):
  File "chrome://sites/content/dinamicTree.js", line 150, in 

[2015-05-05 16:24:24,272] [ERROR] console-logger: TypeError: properties is undefined (2) in chrome://sites/content/dinamicTree.js:150
Traceback (most recent call last):
  File "chrome://sites/content/dinamicTree.js", line 150, in 

[2015-05-05 16:24:24,618] [ERROR] console-logger: TypeError: properties is undefined (2) in chrome://sites/content/dinamicTree.js:150
Traceback (most recent call last):
  File "chrome://sites/content/dinamicTree.js", line 150, in 

[2015-05-05 16:24:25,378] [ERROR] console-logger: TypeError: properties is undefined (2) in chrome://sites/content/dinamicTree.js:148
Traceback (most recent call last):
  File "chrome://sites/content/dinamicTree.js", line 148, in 

[2015-05-05 16:24:30,289] [WARNING] koScintillaSchemeService: applyScheme:: no indicator for name 'collab_remote_cursor_4'
[2015-05-05 16:24:30,289] [WARNING] koScintillaSchemeService: applyScheme:: no indicator for name 'collab_remote_cursor_5'
[2015-05-05 16:24:30,289] [WARNING] koScintillaSchemeService: applyScheme:: no indicator for name 'collab_remote_cursor_2'
[2015-05-05 16:24:30,289] [WARNING] koScintillaSchemeService: applyScheme:: no indicator for name 'collab_remote_cursor_3'
[2015-05-05 16:24:30,290] [WARNING] koScintillaSchemeService: applyScheme:: no indicator for name 'collab_remote_cursor_1'
[2015-05-05 16:24:30,290] [WARNING] koScintillaSchemeService: applyScheme:: no indicator for name 'collab_remote_change'
[2015-05-05 16:24:30,290] [WARNING] koScintillaSchemeService: applyScheme:: no indicator for name 'collab_local_change'
[2015-05-05 16:24:30,365] [ERROR] console-logger: TypeError: properties is undefined (2) in chrome://sites/content/dinamicTree.js:148
Traceback (most recent call last):
  File "chrome://sites/content/dinamicTree.js", line 148, in 

[2015-05-05 16:25:01,676] [WARNING] console-logger: Key event not available on GTK2: key="u" modifiers="control,shift" (1) in chrome://komodo/content/tail/tail.xul:0
[2015-05-05 16:25:01,677] [WARNING] console-logger: Key event not available on some keyboard layouts: key="v" modifiers="control,alt" (1) in chrome://komodo/content/tail/tail.xul:0

Regards

**Comment by cgchoffman** [comment]
Odd. whats the full path you're trying to open?

**Comment by johnnyamr** [comment]
SFTP connection is in an account set up by ip number and leaving as default path / home / httpd. A path would be something like ip_number / directory (for example http://11.111.111.111/wwwdir/), nothing strange. As I say, the version 8 of komodo works seamlessly with the same settings.

**Comment by cgchoffman** [comment]
That doesn't look right. Can you attach a screen shot of your settings for the server?

**Comment by johnnyamr** [comment]
OK.
![server_config](https://cloud.githubusercontent.com/assets/12254129/7522957/e6351c5a-f4ce-11e4-84b4-24b0821dc6fb.gif)

**Comment by Naatan** [comment]
Seems its having issues decoding the SFTP responses. Not sure what about the OS upgrade could have cost this.

Does using SCP instead work?

**Comment by johnnyamr** [comment]
Yes, I have no problem with sftp to work with other accounts on different machines and even with the same account using komodo version 8.5.4, build 14424, I work well. I think I will continue using version 8.5 waiting for future updates and any possible solution.

**Comment by Naatan** [comment]
Unfortunately as we are unable to reproduce this issue we cannot fix it. It's also not clear to me whether this is an issue with Komodo or your system config, or that of the system you are connecting to.

Closing this until it becomes apparent that the issue is with Komodo (ie. more people with the issue).

**Comment by kodhtml** [comment]
@Naatan I have similar issue. I can provide you access to sftp account to test it. On 8.5 everything works well, but not any more on komode 9.

**Comment by johnnyamr** [comment]
Hello, solves the problem by changing the account settings in the server type, SCP by SFTP. Komodo Edit 9.3.2 tested. Regards.

**Comment by Naatan** [comment]
@kodhtml if you are willing to provide ftp server access to a server which reproduces the problem it'd be much appreciated. You can email the details to support@activestate.com, please include a link to this ticket.

**Comment by kodhtml** [comment]
@johnnyamr it doesn't work for me.
@Naatan ok, sent.

**Comment by Naatan** [comment]
Thank you @kodhtml. Oddly enough it seems to work fine for me using SFTP.

What platform/os are you on?

**Comment by kodhtml** [comment]
I'm using Win 10 version 1511, compilation 10586.104 - all updates. I made additional tests (various types, e.g. with/without antivirus, different computer with Win 7, but on the same internet connection) and:
- FTP works well
- SFTP throw this message:

> Error: '[Exception... "Component returned failure code: 0x80004005 (NS_ERROR_FAILURE) [koIRemoteConnectionService.getConnection2]"  nsresult: "0x80004005 (NS_ERROR_FAILURE)"  location: "JS frame :: chrome://komodo/content/dialogs/filebrowser/filebrowser.js :: connectToServer :: line 1592"  data: no]'

Let me know if you need additional information or test.

**Comment by Naatan** [comment]
Thanks @kodhtml, @cgchoffman would you mind digging on this when you're back? FTP details are on OTRS ticket #46086

**Comment by cgchoffman** [comment]
@naatan, @kodhtml, I tested this on OSX and cannot reproduce it.
Open *View > tabs & sidebars > Console* and paste/execute the following code then reproduce the issue:
```
ko.logging.getLogger('filepicker.remote').setLevel(10);
```

**Comment by joho1968** [comment]
I just started seeing this as well, remote directory SFTP. Running KomodoIDE 9.3 under Linux. Never seen this particular error before.

**Comment by Naatan** [comment]
@cgchoffman could you test it on Windows?

**Comment by cgchoffman** [comment]
I use remote sftp drives on Windows all the time.  I have no issues in komodo.

**Comment by Naatan** [comment]
Could you test it with the server he provided?

**Comment by cgchoffman** [comment]
...Why did I test this on OSX any way?  I get the same error as @kodhtml in this comment: https://github.com/Komodo/KomodoEdit/issues/252#issuecomment-183247521

@joho1968 your issue is different as far as I can tell.  The issue here is that we cannot connect to the server at all.  You have connected with out issue it looks like, you simply can't save one of your files.  Please open a new issue report.

**Comment by cgchoffman** [comment]
So i think I've traced this down to Paramiko, the Python package we use for remote connections.  I believe it's related to this ticket: https://github.com/paramiko/paramiko/issues/353.  I just tried applying the patch to test but I need to rebuild Komodo and that will take too long...I need to go to French class.

See line 150 in the attached logs to see WHY I think it's paramiko and that particular error.

[remote_connections_debug.txt](https://github.com/Komodo/KomodoEdit/files/141743/remote_connections_debug.txt)

**Comment by Naatan** [comment]
Nice digging @cgchoffman !

**Comment by cgchoffman** [comment]
I can resolve this issue by pulling in `Paramiko 1.15.4` and dropping it in Komodo.  If anyone following this ticket wants to do that themselves, it's pretty straight forward.  It's not thoroughly tested though so do it at your own risk and i obviously can't recommend you do it.  It would be up to you.  I can tell you how to do it.

Paramiko 1.16.0 (latest) has this issue which breaks on Windows 32 Python builds: https://github.com/paramiko/paramiko/issues/613

As long as this stays on Komodo 10 (sorry i could have sworn this was a 9.3.x bug otherwise I wouldn't have worked on it) I'm fine with dropping in the 1.15.4 into master or we can wait to see if 1.16.1 comes out before 10 is released.

**Comment by Naatan** [comment]
It's too big a change for a minor release imo, unless you can cherry pick the specific fix.

**Comment by cgchoffman** [comment]
@kodhtml did you remove the test account you setup for us?  Any chance you could add it back for a day (monday)?  I can't repro this without your server and I want to double check my fix before I commit it.

**Comment by kodhtml** [comment]
@cgchoffman no, there was a problem with server, but now it works.

**Comment by cgchoffman** [comment]
Thanks for fixing that @kodhtml.  After retesting the fix, I'm getting conflicting results.  v1.15.4 now returns the exact same error as our current version which is 1.15.1.  I"m getting the same result when I try the latest tag v1.16.

We'll have to visit this later.  

Since so few people are hitting this I wonder if it might be a server config issue?  Something to check out on your end @kodhtml.

**Comment by blah-dj** [comment]
@cgchoffman  can you let me know the error on server? I am seeing same error now. It was working fine till few days ago but today i am seeing same error with sftp/ftp or even scp.

**Comment by cgchoffman** [comment]
@blah-dj, it's the error in @kodhtml.  
https://github.com/Komodo/KomodoEdit/issues/252#issuecomment-183247521

@blah-dj I'd be looking at your server if you have only just recently started having issues.  That's assuming you didn't update Komodo in those days.

**Comment by blah-dj** [comment]
Its a private server (no internet access). I do a webex and show you the issue. I have the most recent komodo edit version
Komodo Edit, version 9.3.2, build 16460, platform win32-x86. Built on Thu Dec 10 22:21:38 2015.
on windows 7. 

it was working fine few days back.

**Comment by cgchoffman** [comment]
Don't really have to do one on one trouble shooting.  We're in full development mode.  Could start by providing the actually error you're getting and back track through any changes you might have made recently.  As I said, if you haven't changed Komodo then something changed in your system.

**Comment by cgchoffman** [comment]
More notes:
I've finally been able to get this to work
- paramiko 1.15.4
- PuTTY  0.67
- Apply this patch: 
[252-patch.txt](https://github.com/Komodo/KomodoEdit/files/241370/252-patch.txt)

**Comment by cgchoffman** [comment]
Note: we're too far into 10 dev to add this.  Moving to 10.1, very sorry for the continued delay everyone who is waiting on this.  Again I can assist in applying the above fix if you really need it.  It's about 15 minutes or less if you want to.

**Comment by blah-dj** [comment]
Can you send me the instructions to add this?

Thanks,
Divya Jain

**Comment by cgchoffman** [comment]
- Git clone paramiko into a temp folderit
`git clone https://github.com/paramiko/paramiko.git`
- enter the folder
`cd paramiko`
- in that folder checkout the v1.15.4 tag
`git checkout v1.15.4`
- rename `KOMODO_INS_DIR\Komodo IDE 9.3.2\lib\mozilla\python\komodo\paramiko` to move it aside
- create a new `paramiko` folder
- copy the contents of the `paramiko/paramiko` folder into `KOMODO_INS_DIR\Komodo IDE 9.3.2\lib\mozilla\python\komodo\paramiko`
- rename the following file `KOMODO_INS_DIR\lib\mozilla\python\komodo\remotefilelib.py`
- copy this file in it's place
[remotefilelib.py.txt](https://github.com/Komodo/KomodoEdit/files/241536/remotefilelib.py.txt)
- make sure you rename it with the correct file ext. (ie. `.py`)

**Comment by tnfno** [comment]
I have a similar error on KO Edit version 10.0.1, build 17276 on a Windows 10 build 14366 PC:

`Error: '[Exception... "Component returned failure code: 0x80004005 (NS_ERROR_FAILURE) [koIRemoteConnectionService.getConnection2]"  nsresult: "0x80004005 (NS_ERROR_FAILURE)"  location: "JS frame :: chrome://komodo/content/dialogs/filebrowser/filebrowser.js :: connectToServer :: line 1592"  data: no]'`

Was this supposed to be fixed in 10 or should I try the same paramiko patch ?

**Comment by cgchoffman** [comment]
@tnfno milestone is set for 10.1.  This hasn't been added yet but thanks for updating your comment.

**Comment by cgchoffman** [comment]
My apologize to anyone waiting on this.  I dropped the ball and didn't implement this in time for 10.1.
